CoreFeatures.md

```diff
@@ -0,0 +1,181 @@
+
+NOTE: The primitive type `INT96` is deprecated so it is intentionally not listed
+here.
+
+#### Logical types
+
+The [logical type](LogicalTypes.md)s are practically annotations helping to
+understand the related primitive type (or structure). Originally we have had
+the `ConvertedType` enum in the thrift file representing all the possible
+logical types. After a while we realized it is hard to extend and so introduced
+the `LogicalType` union. For backward compatibility reasons we allow to use the
+old `ConvertedType` values according to the specified rules but we expect that
+the logical types in the file schema are defined with `LogicalType` objects.
+
+The following LogicalTypes are supported:
+* `STRING`
+* `MAP`
+* `LIST`
+* `ENUM`
+* `DECIMAL` (for which primitives?)
+* `DATE`
+* `TIME`: **(Which unit, utc?)**
+* `TIMESTAMP`: **(Which unit, utc?)**
+* `INTEGER`: (all bitwidth 8, 16, 32, 64) **(unsigned?)**
+* `UNKNOWN` **(?)**
+* `JSON` **(?)**
+* `BSON` **(?)**
+* `UUID` **(?)**
+
+NOTE: The old ConvertedType `INTERVAL` has no representation in LogicalTypes.
+This is becasue `INTERVAL` is deprecated so we do not include it in this list.
+
+### Encodings
+
+The following encodings are supported:
+* [PLAIN](Encodings.md#plain-plain--0)  
+  parquet-mr: Basically all value types are written in this encoding in case of
+  V1 pages
+* [PLAIN\_DICTIONARY](Encodings.md#dictionary-encoding-plain_dictionary--2-and-rle_dictionary--8)
+  **(?)**  
+  parquet-mr: As per the spec this encoding is deprecated while we still use it
+  for V1 page dictionaries.
+* [RLE](Encodings.md#run-length-encoding--bit-packing-hybrid-rle--3)  
+  parquet-mr: Used for both V1 and V2 pages to encode RL and DL and for BOOLEAN
+  values in case of V2 pages
+* [DELTA\_BINARY\_PACKED](Encodings.md#delta-encoding-delta_binary_packed--5)
+  **(?)**  
+  parquet-mr: Used for V2 pages to encode INT32 and INT64 values.
+* [DELTA\_LENGTH\_BYTE\_ARRAY](Encodings.md#delta-length-byte-array-delta_length_byte_array--6)
+  **(?)**  
+  parquet-mr: Not used directly
+* [DELTA\_BYTE\_ARRAY](Encodings.md#delta-strings-delta_byte_array--7)
+  **(?)**  
+  parquet-mr: Used for V2 pages to encode BYTE\_ARRAY and
+  FIXED\_LEN\_BYTE\_ARRAY values
+* [RLE\_DICTIONARY](Encodings.md#dictionary-encoding-plain_dictionary--2-and-rle_dictionary--8)
+  **(?)**  
+  parquet-mr: Used for V2 page dictionaries
+* [BYTE\_STREAM\_SPLIT](Encodings.md#byte-stream-split-byte_stream_split--9)
+  **(?)**  
+  parquet-mr: Not used by default; can be used only via explicit configuration
+
+NOTE: [BIT\_PACKED](Encodings.md#bit-packed-deprecated-bit_packed--4) is
+deprecated and not used directly (boolean values are encoded with this under
+PLAIN) so not included in this list.
+
+**TODO**: In parquet-mr dictionary encoding is not enabled for
+FIXED\_LEN\_BYTE\_ARRAY in case of writing V1 pages. I don't know the reason
+behind. Any experience/idea about this from other implementations?
```

Comment:
Impala does this for decimal values. I think this is fine and makes sense to do. E.g. in the TPC-H and probably TPC-DS data sets there are some cardinality decimal columns

[localhost.EXAMPLE.COM:21050] default> show column stats tpch_parquet.lineitem;
Query: show column stats tpch_parquet.lineitem
+-----------------+---------------+------------------+--------+----------+---------------+--------+---------+
| Column          | Type          | #Distinct Values | #Nulls | Max Size | Avg Size      | #Trues | #Falses |
+-----------------+---------------+------------------+--------+----------+---------------+--------+---------+
| l_orderkey      | BIGINT        | 1563438          | 0      | 8        | 8.0           | -1     | -1      |
| l_partkey       | BIGINT        | 200516           | 0      | 8        | 8.0           | -1     | -1      |
| l_suppkey       | BIGINT        | 9712             | 0      | 8        | 8.0           | -1     | -1      |
| l_linenumber    | INT           | 7                | 0      | 4        | 4.0           | -1     | -1      |
| l_quantity      | DECIMAL(12,2) | 51               | 0      | 8        | 8.0           | -1     | -1      |
| l_extendedprice | DECIMAL(12,2) | 868550           | 0      | 8        | 8.0           | -1     | -1      |
| l_discount      | DECIMAL(12,2) | 11               | 0      | 8        | 8.0           | -1     | -1      |
| l_tax           | DECIMAL(12,2) | 9                | 0      | 8        | 8.0           | -1     | -1      |
| l_returnflag    | STRING        | 3                | 0      | 1        | 1.0           | -1     | -1      |
| l_linestatus    | STRING        | 2                | 0      | 1        | 1.0           | -1     | -1      |
| l_shipdate      | STRING        | 2629             | 0      | 10       | 10.0          | -1     | -1      |
| l_commitdate    | STRING        | 2559             | 0      | 10       | 10.0          | -1     | -1      |
| l_receiptdate   | STRING        | 2658             | 0      | 10       | 10.0          | -1     | -1      |
| l_shipinstruct  | STRING        | 4                | 0      | 17       | 11.9986381531 | -1     | -1      |
| l_shipmode      | STRING        | 7                | 0      | 7        | 4.28530454636 | -1     | -1      |
| l_comment       | STRING        | 4652621          | 0      | 43       | 26.4941692352 | -1     | -1      |
+-----------------+---------------+------------------+--------+----------+---------------+--------+---------+

Reply:
I believe Arrow/Parquet CPP also does dictionary encoding here (would need to double check).

Reply:
Thanks for the info. I'll remove this TODO.